tests/test_rpub.py

```diff
@@ -0,0 +1,837 @@
+            except RuntimeError as e:
+                print(i, e)
+                raise
+
+    # test that invalid unicode in message header keys is detected
+    corrupted_data = bytearray(journal_data).replace(b"header_0",
+                                                     b"\xc3\x28\xa0\xa1\xf0\x28\x8c\xbc")
+    with open(journal_path, "wb") as journal_file:
+        journal_file.write(corrupted_data)
+    with pytest.raises(RuntimeError):
+        try:
+            j = PublicationJournal(journal_path)
+        except RuntimeError as e:
+            print("invalid unicode message header", e)
+            raise
+
+
+def test_journal_restore_state_truncated(tmpdir):
+    journal_path = tmpdir.join("journal")
+    j = PublicationJournal(journal_path)
+
+    headers = [(b"header_0", b"value_0"), (b"header_1", b"value_1")]
+    for i in range(0, 10):
+        message = f"message {i}".encode("utf-8")
+        if (i % 3) == 0:
+            j.queue_message(message, headers=headers)
+        else:
+            j.queue_message(message)
+
+    for i in range(0, 10):
+        m = j.get_next_message_to_send()
+        if (i % 2) == 0:
+            j.mark_message_sent(m[0])
+
+    del j
+    # capture the raw journal for later manipulation
+    with open(journal_path, "rb") as journal_file:
+        journal_data = journal_file.read()
+
+    # these are the offsets of the ends of the records
+    # if the file format or test data is changed, these must be updated
+    # truncating the file between records is valid/undetectable, so we should not test these
+    valid_break_points = [119, 176, 233, 352, 409, 466, 585, 642, 699, 818,
+                          850, 882, 914, 946, 978]
+    # test that truncation at any byte (in a record) is detected
+    for i in range(1, len(journal_data)):
+        if i in valid_break_points:
+            continue
+        with open(journal_path, "wb") as journal_file:
+            journal_file.write(journal_data[0:i])
+        with pytest.raises(RuntimeError):
+            try:
+                j = PublicationJournal(journal_path)
+            except RuntimeError as e:
+                print(i, e)
+                raise
+            print(f"Failure: Truncation at byte {i} should be detected")
+
+
+def test_journal_restore_duplicate_sequence_number_to_send(tmpdir):
+    journal_path = tmpdir.join("journal")
+    j = PublicationJournal(journal_path)
+
+    # directly use internal recording function to generate records with valid checksums
+    # but dubious meanings
+    test_message = b"data"
+    rbody = BytesIO()
+    rbody.write(PublicationJournal.encode_int(0))  # sequence number
+    rbody.write(PublicationJournal.encode_int(len(test_message)))  # data length
+    rbody.write(test_message)  # data
+    rbody.write(PublicationJournal.encode_int(0))  # no headers
+    j._write_record(PublicationJournal.msg_record_type, rbody.getvalue())
+    # record the same message again, duplicating the sequence number
```

Comment:
True, and that was more generally what the test was intended for, since any duplicate sequence number is invalid. Originally I was being lazy and not encoding a seocnd record, but I've now extended the test to do that, too.